src/lib.rs

```diff
@@ -145,6 +156,60 @@ impl<'data> Iterator for DataFormatDescriptorIterator<'data> {
     }
 }
 
+struct KeyValueDataIterator<'data> {
+    data: &'data [u8],
+}
+
+impl<'data> Iterator for KeyValueDataIterator<'data> {
+    type Item = (&'data str, &'data [u8]);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut offset = 0;
+
+        loop {
+            let length = bytes_to_u32(self.data, &mut offset).ok()?;
+
+            let start_offset = offset;
+
+            offset += length as usize;
+
+            // Ensure that we're 4-byte aligned
+            if offset % 4 != 0 {
+                offset += 4 - (offset % 4);
+            }
+
+            let key_and_value = match self.data.get(start_offset..start_offset + length as usize) {
+                Some(key_and_value) => key_and_value,
+                None => continue,
+            };
+
+            // The key is terminated with a NUL character.
+            let key_end_index = match key_and_value.iter().position(|&c| c == b'\0') {
+                Some(key_and_value) => key_and_value,
+                None => continue,
+            };
+
+            let key = &key_and_value[..key_end_index];
+            let value = &key_and_value[key_end_index + 1..];
+
+            let key = match std::str::from_utf8(key) {
+                Ok(key) => key,
+                Err(_) => continue,
+            };
+
+            self.data = match self.data.get(offset..) {
+                Some(data) => data,
```

Comment:
Do we need to set `offset` back to 0 at this point? Do we actually want `offset` to outlive a single loop iteration at all?

Reply:
If think so, because we want the offset to stay the same over `continue`s.

Reply:
Oh, I see, we don't need to reset `offset` here because once we get this far we're guaranteed to `return` and hence it no longer matters.